test/extended/devices/recovery.go

```diff
@@ -0,0 +1,416 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	kubeletdevicepluginv1beta1 "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/openshift/origin/test/extended/util/image"
+)
+
+const (
+	workloadCommand               = "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo devices=$devs && sleep 24h" // sleep forever in the test timescale
+	setupRegistrationCommandMCD   = "mkdir -p /rootfs/var/lib/kubelet/device-plugins/sample && touch /rootfs/var/lib/kubelet/device-plugins/sample/registration"
+	unblockRegistrationCommandMCD = "rm /rootfs/var/lib/kubelet/device-plugins/sample/registration"
+	rebootNodeCommandMCD          = "chroot /rootfs systemctl reboot"
+
+	sampleDeviceResourceName            = "example.com/resource"
+	sampleDevicePluginPodName           = "sample-device-plugin"
+	sampleDeviceEnvVarNamePluginSockDir = "PLUGIN_SOCK_DIR"
+
+	sampleDeviceEnvVarNameControlFile = "REGISTER_CONTROL_FILE"
+	sampleDevicePluginControlFilePath = "/var/lib/kubelet/device-plugins/sample/registration"
+
+	sampleDevicePluginPodDSYAML = "test/e2e/testing-manifests/sample-device-plugin/sample-device-plugin-control-registration.yaml"
+
+	targetNodeEnvVar = "DEVICE_RECOVERY_TARGET_NODE"
+
+	rebootCooldownEnvVar  = "REBOOT_COOLDOWN_TIME"
+	rebootCooldownDefault = 1 * time.Minute
+)
+
+var (
+	sampleDevicePluginPodLabel = map[string]string{
+		"device-recovery-test-pod": "device-plugin",
+	}
+	workloadPodLabel = map[string]string{
+		"device-recovery-test-pod": "workload",
+	}
+)
+
+var _ = g.Describe("[sig-node][Serial][Slow][Disruptive][Feature:DeviceManager] Device management tests", func() {
+	defer g.GinkgoRecover()
+
+	var (
+		oc = exutil.NewCLIWithPodSecurityLevel("devices", admissionapi.LevelPrivileged)
+	)
+
+	g.It("Verify that pods requesting devices are correctly recovered on node restart", func() {
+		oc := oc.AsAdmin()
+
+		targetNode := os.Getenv(targetNodeEnvVar)
+		if targetNode == "" {
+			e2eskipper.Skipf("Need an explicit target node name, got none")
+		}
+		framework.Logf("target node name: %q", targetNode)
+
+		var err error
+		rebootCooldownTime := rebootCooldownDefault
+		if val, ok := os.LookupEnv(rebootCooldownEnvVar); ok {
+			rebootCooldownTime, err = time.ParseDuration(val)
+			o.Expect(err).NotTo(o.HaveOccurred(), "error creating workload pod")
+		}
+		framework.Logf("reboot cooldown time: %v", rebootCooldownTime)
+
+		fwk := oc.KubeFramework()
+		namespace := fwk.Namespace.Name // shortcut
+		k8sCli := oc.KubeClient()
+
+		// phase1: figure out the node and prepare it
+		node, err := k8sCli.CoreV1().Nodes().Get(context.Background(), targetNode, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred(), "error getting the target node %q", targetNode)
```

Comment:
NIT: usually we use `ToNot` which is more readable
So we read:
Except error _to not_ have occurred other 
instead of:
Except error _not to_ have occurred

Reply:
Absolutely, no nit at all! thanks for pointing this out. Fixed in the last upload